### ISSUE TITLE
CRASH (any WOW64 app) on AMD Ryzen in drlibc_x86.asm:566 WOW64 mode switch

### ISSUE THREAD

**Issue by justintaft** [DynamoRIO/dynamorio#4091]
**Describe the bug**

bin32\ddrun.exe crashes most of the time when operating on 32 bit applications. Maybe one out of 100 times notepad.exe launches. In my setup, Windows 10 is being virtualized by KVM. AMD 3900x Processor is being used.  Older versions of DynamoRio has the same issue.  bin64\drrun.exe works fine for 64 bit applications.

```
C:\Users\justin\Downloads\DynamoRIO-Windows-7.91.18301-1\DynamoRIO-Windows-7.91.18301-1>bin32\drrun.exe -- C:\windows\syswow64\notepad.exe
```

![image](https://user-images.githubusercontent.com/8340498/74109412-38775400-4b51-11ea-94e8-271694ac1a3a.png)

**To Reproduce**
1. Download Window's cronbuild-7.91.18301A from https://github.com/DynamoRIO/dynamorio/releases . 
2. Extract and cd to DynamoRIO folder.
3. Run the command `bin32\drrun.exe -- C:\windows\syswow64\notepad.exe`
4. Note the application likely crashes.
5. Repeat step 3  around 100 times, and notice notepad launches.

Please also answer these questions:
 - What happens when you run without any client?

No client is being used. Clients do work, but again need to run 100 times or so to get the client to run. I

 - What happens when you run with debug build ("-debug" flag to drrun/drconfig/drinject)?

Same issue occurs, but with more detailed error message. When notepad manages to start, it takes a longer time for the process to open.

Image showing the error message:

![image](https://user-images.githubusercontent.com/8340498/74109498-ebe04880-4b51-11ea-9bf7-8817ca4a4b04.png)

Image showing notepad.exe does launch once in a while:

![image](https://user-images.githubusercontent.com/8340498/74109515-1f22d780-4b52-11ea-9cf2-ac2722c341b0.png)

**Expected behavior**

Notepad to open and run.

**Screenshots or Pasted Text**

Following is StackTrace in WinDBG. The above reproduction steps is for starting in windows without
```
ModLoad: 75ab0000 75ad5000   C:\Windows\SysWOW64\IMM32.DLL
ModLoad: 14460000 14909000   C:\Users\justin\Downloads\DynamoRIO-Windows-7.91.18301-1\DynamoRIO-Windows-7.91.18301-1\lib32\debug\dynamorio.dll
(1348.2a8): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
eax=00000000 ebx=00fbf490 ecx=b58d14d4 edx=00000000 esi=2190b860 edi=00fbf540
eip=14775eab esp=00fbf4b0 ebp=00fbf4e4 iopl=0         nv up ei pl nz ac po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010212
dynamorio!switch_modes_and_call+0x4e:
14775eab 5b              pop     ebx
1:002> k
 # ChildEBP RetAddr  
00 00fbf4e4 146fe909 dynamorio!switch_modes_and_call+0x4e [C:\projects\dynamorio\build\build_debug-32\core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s @ 1961] 
01 00fbf510 146fed98 dynamorio!os_take_over_wow64_extra+0x49 [c:\projects\dynamorio\core\win32\os.c @ 2433] 
02 00fbfa30 146f1928 dynamorio!os_take_over_thread+0x2b8 [c:\projects\dynamorio\core\win32\os.c @ 2571] 
03 00fbfa70 1446fc58 dynamorio!os_take_over_all_unknown_threads+0x1f8 [c:\projects\dynamorio\core\win32\os.c @ 2707] 
04 00fbfaa4 146d7322 dynamorio!dynamorio_take_over_threads+0xf8 [c:\projects\dynamorio\core\dynamo.c @ 2904] 
05 00fbfb10 146d7aa8 dynamorio!auto_setup+0x202 [c:\projects\dynamorio\core\arch\x86_code.c @ 198] 
06 00000000 00000000 dynamorio!dynamo_auto_start+0x8 [C:\projects\dynamorio\build\build_debug-32\core\CMakeFiles\dynamorio.dir\arch\x86\x86.asm.obj.s @ 1600] 
```

```
1:003> .exr -1
ExceptionAddress: 70afe90b (dynamorio!switch_modes_and_call+0x0000004e)
   ExceptionCode: c0000005 (Access violation)
  ExceptionFlags: 00000000
NumberParameters: 2
   Parameter[0]: 00000003
   Parameter[1]: 00000000
Attempt to write to address 00000000
```
```
0:000> !address esp

                                     
Mapping file section regions...
Mapping module regions...
Mapping PEB regions...
Mapping TEB and stack regions...
Mapping heap regions...
Mapping page heap regions...
Mapping other regions...
Mapping stack trace database regions...
Mapping activation context regions...

Usage:                  Stack
Base Address:           00fff000
End Address:            01010000
Region Size:            00011000 (  68.000 kB)
State:                  00001000          MEM_COMMIT
Protect:                00000004          PAGE_READWRITE
Type:                   00020000          MEM_PRIVATE
Allocation Base:        00fd0000
Allocation Protect:     00000004          PAGE_READWRITE
More info:              ~0k


Content source: 1 (target), length: bd0
```

**Versions**
 - What version of DynamoRIO are you using?
Windows-7.91.18301

 - Does the latest build from
https://github.com/DynamoRIO/dynamorio/wiki/Latest-Build solve the problem?
No

- What operating system version are you running on? ("Windows 10" is *not* sufficient: give the release number.)

Windows 10 1809 (Build 17763.973) is being used. I am virtualizing  Windows through  KVM.

- Is your application 32-bit or 64-bit?

32 bit

**Additional context**
None

**Comment by derekbruening** [comment]
Thank you for the detailed error report.  Unfortunately (as expected) I can't reproduce this on my machine: WOW64 notepad runs fine 10x (I would expect Nx) in a row under 7.91.18301 on Win10 1909 on physical hardware.  That's 1909 vs your 1809 but I'm pretty sure any prior failure on notepad on a prior Win10 version would have been noticed since it is a typical test app.  I would guess there is something particular about your setup: maybe the KVM.  Have you seen this on any other machine?

Unfortunately os_take_over_wow64_extra() is complex, and the Win10 system threads bring it into play in every app nowadays.  If you are comfortable with windbg and low-level debugging maybe you could dig further into what is going wrong?  The windbg data shown so far is confusing: if the `pop ebx` is what faulted why would it be a fault on an `Attempt to write to address 00000000`.  Something is off there.

**Comment by justintaft** [comment]
Thank you for trying to re-produce so quick.  When I have time I'll setup booting into windows directly and giving it a go. The interesting thing to is

```
   Parameter[0]: 00000003
```
Is this normal? I thought normal values are  Read (0), Write (1), Execute (8) 
 (mentioned in article on https://channel9.msdn.com/Shows/Inside/C0000005)

**Comment by derekbruening** [comment]
> The other interesting thing is the exception parameter 0.
> 
> ```
>    Parameter[0]: 00000003
> ```
> 
> Is this normal? I thought normal values are Read (0), Write (1), Execute (8)
> (mentioned in article on https://channel9.msdn.com/Shows/Inside/C0000005)

Wow, no, I have never seen that before.  Every source I've seen says it will be 0, 1, or 8.  First thought is whether that record is corrupted: though it says it's a first-chance.  Hmm.

**Comment by justintaft** [comment]
I also compiled from source, still ran into the same error.  

I modified the assembly file to place an `int 3` instruction before `pop ebx`.  Oddly enough when I run the application under WinDbG with the breakpoint, the breakpoint is hit a few times.  I press 'g' to continue, and I  haven't seen the Access Violation being thrown on `pop ebx`.   Access violation in 
`dynamorio!safe_read_asm_mid` is returned, but I read issue #3846 this was expected.  

I'm a bit confused why this works. If you feel this is not related to DynamoRio, I will limit updates until I find more concrete explanation about what's going on.

```
Microsoft (R) Windows Debugger Version 10.0.19528.1000 X86
Copyright (c) Microsoft Corporation. All rights reserved.

CommandLine: C:\Users\justin\Desktop\dynamorio\build\bin32\drrun.exe --  C:\windows\syswow64\notepad.exe
Starting directory: C:\Users\justin\Desktop\dynamorio\build\bin32

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is: 
ModLoad: 00ca0000 00cd8000   drrun.exe
ModLoad: 77b10000 77cac000   ntdll.dll
ModLoad: 77840000 77920000   C:\Windows\SysWOW64\KERNEL32.DLL
ModLoad: 76710000 76909000   C:\Windows\SysWOW64\KERNELBASE.dll
ModLoad: 0fe60000 0fe99000   C:\Users\justin\Desktop\dynamorio\build\bin32\drconfiglib.dll
ModLoad: 77620000 7769e000   C:\Windows\SysWOW64\ADVAPI32.dll
ModLoad: 75190000 75250000   C:\Windows\SysWOW64\msvcrt.dll
ModLoad: 75c10000 75c89000   C:\Windows\SysWOW64\sechost.dll
ModLoad: 76650000 7670f000   C:\Windows\SysWOW64\RPCRT4.dll
ModLoad: 75170000 75190000   C:\Windows\SysWOW64\SspiCli.dll
ModLoad: 75160000 7516a000   C:\Windows\SysWOW64\CRYPTBASE.dll
ModLoad: 76970000 769d2000   C:\Windows\SysWOW64\bcryptPrimitives.dll
ModLoad: 0f5a0000 0f62a000   C:\Users\justin\Desktop\dynamorio\build\bin32\drinjectlib.dll
(36c8.2068): Break instruction exception - code 80000003 (first chance)
eax=00000000 ebx=00f1b000 ecx=2e6b0000 edx=00000000 esi=01171d10 edi=77b237ec
eip=77bbf146 esp=0113f51c ebp=0113f548 iopl=0         nv up ei pl zr na pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000246
ntdll!LdrpDoDebuggerBreak+0x2b:
77bbf146 cc              int     3
0:000> g

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is: 
ModLoad: 012f0000 0132f000   notepad.exe
ModLoad: 77b10000 77cac000   ntdll.dll
ModLoad: 77840000 77920000   C:\Windows\SysWOW64\KERNEL32.DLL
ModLoad: 76710000 76909000   C:\Windows\SysWOW64\KERNELBASE.dll
ModLoad: 77620000 7769e000   C:\Windows\SysWOW64\ADVAPI32.dll
ModLoad: 75190000 75250000   C:\Windows\SysWOW64\msvcrt.dll
ModLoad: 75c10000 75c89000   C:\Windows\SysWOW64\sechost.dll
ModLoad: 76650000 7670f000   C:\Windows\SysWOW64\RPCRT4.dll
ModLoad: 75170000 75190000   C:\Windows\SysWOW64\SspiCli.dll
ModLoad: 75160000 7516a000   C:\Windows\SysWOW64\CRYPTBASE.dll
ModLoad: 76970000 769d2000   C:\Windows\SysWOW64\bcryptPrimitives.dll
ModLoad: 75ae0000 75b03000   C:\Windows\SysWOW64\GDI32.dll
ModLoad: 76490000 765f6000   C:\Windows\SysWOW64\gdi32full.dll
ModLoad: 75c90000 75d10000   C:\Windows\SysWOW64\msvcp_win.dll
ModLoad: 75970000 75a92000   C:\Windows\SysWOW64\ucrtbase.dll
ModLoad: 757d0000 7596a000   C:\Windows\SysWOW64\USER32.dll
ModLoad: 75580000 75597000   C:\Windows\SysWOW64\win32u.dll
ModLoad: 75300000 75578000   C:\Windows\SysWOW64\combase.dll
ModLoad: 756d0000 757ca000   C:\Windows\SysWOW64\COMDLG32.dll
ModLoad: 755a0000 75629000   C:\Windows\SysWOW64\shcore.dll
ModLoad: 752b0000 752f4000   C:\Windows\SysWOW64\SHLWAPI.dll
ModLoad: 75d10000 76263000   C:\Windows\SysWOW64\SHELL32.dll
ModLoad: 769e0000 76a1b000   C:\Windows\SysWOW64\cfgmgr32.dll
ModLoad: 77020000 7761e000   C:\Windows\SysWOW64\windows.storage.dll
ModLoad: 776a0000 776bc000   C:\Windows\SysWOW64\profapi.dll
ModLoad: 76f60000 76fb4000   C:\Windows\SysWOW64\powrprof.dll
ModLoad: 76960000 7696f000   C:\Windows\SysWOW64\kernel.appcore.dll
ModLoad: 76270000 76282000   C:\Windows\SysWOW64\cryptsp.dll
ModLoad: 72fa0000 731af000   C:\Windows\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.17763.973_none_4d5f7c671b9e1487\COMCTL32.dll
ModLoad: 74fe0000 7515f000   C:\windows\SysWOW64\PROPSYS.dll
ModLoad: 75630000 756c6000   C:\Windows\SysWOW64\OLEAUT32.dll
ModLoad: 4c5f0000 4c65b000   C:\windows\SysWOW64\WINSPOOL.DRV
ModLoad: 76920000 76939000   C:\Windows\SysWOW64\bcrypt.dll
ModLoad: 73320000 734de000   C:\windows\SysWOW64\urlmon.dll
ModLoad: 74880000 748b3000   C:\windows\SysWOW64\IPHLPAPI.DLL
ModLoad: 714d0000 716ff000   C:\windows\SysWOW64\iertutil.dll
(528.1280): Break instruction exception - code 80000003 (first chance)
eax=00000000 ebx=00e24000 ecx=5ccb0000 edx=00000000 esi=03411d58 edi=77b237ec
eip=77bbf146 esp=0107f33c ebp=0107f368 iopl=0         nv up ei pl zr na pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000246
ntdll!LdrpDoDebuggerBreak+0x2b:
77bbf146 cc              int     3
1:002> g
ModLoad: 75ab0000 75ad5000   C:\Windows\SysWOW64\IMM32.DLL
ModLoad: 704e0000 7063b000   C:\Users\justin\Desktop\dynamorio\build\lib32\release\dynamorio.dll
(528.1280): Break instruction exception - code 80000003 (first chance)
eax=00000000 ebx=0107f260 ecx=b58d14d4 edx=00000000 esi=70612266 edi=0107f38c
eip=7058bb1d esp=0107f280 ebp=0107f330 iopl=0         nv up ei pl nz ac po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000212
dynamorio!switch_modes_and_call+0x50:
7058bb1d cc              int     3
1:002> g
(528.1280): Break instruction exception - code 80000003 (first chance)
eax=00000000 ebx=0107f260 ecx=b58d14d4 edx=00000000 esi=70612266 edi=0107f38c
eip=7058bb1d esp=0107f280 ebp=0107f330 iopl=0         nv up ei pl nz ac po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000212
dynamorio!switch_modes_and_call+0x50:
7058bb1d cc              int     3
1:002> g
(528.1280): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
eax=19c8ee28 ebx=77b85c79 ecx=00000001 edx=00000004 esi=fffffffc edi=19c8ee28
eip=7058b953 esp=19c8ed3c ebp=19c8ed54 iopl=0         nv up ei pl nz ac po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010212
dynamorio!safe_read_asm_mid:
7058b953 f3a5            rep movs dword ptr es:[edi],dword ptr [esi]
```

**Comment by justintaft** [comment]
Okay... i found a hacky work-around. I'm able to run notepad everytime with the following patch. I'm just running a long loop before the popping occurs.  It's weird to say the least...maybe it allows other threads/processes to run and somehow fixing w/e weird state is occurring.

```
--- a/core/drlibc/drlibc_x86.asm
+++ b/core/drlibc/drlibc_x86.asm
@@ -561,9 +561,16 @@ smc_transfer_to_64:
         push     offset smc_return_to_32  /* 8-byte push */
         mov      dword ptr [esp + 4], CS32_SELECTOR /* top 4 bytes of prev push */
         jmp      fword ptr [esp]
+
 smc_return_to_32:
-        add      esp, 8          /* clean up far jmp target */
-        pop      ebx             /* restore callee-saved reg */
+
+        add      esp, 08h          /* clean up far jmp target */
+               mov ebx, 800000h
+loop2:
+               dec ebx
+               cmp ebx, 0
+               jne loop2
+               pop      ebx             /* restore callee-saved reg */

```

**Comment by justintaft** [comment]
The above patch doesn't fix the problem completely.  I talked to a friend and they believe some instruction or data caching may be going on. I've been fuzzing with WinAFL and dyno and it's been working for the past days. I decided to fire up multiple instances of AFL, and ran into the same bug again. 


![image](https://user-images.githubusercontent.com/8340498/74626434-a1486880-511d-11ea-9a69-14e67599a1b8.png)

**Comment by davicr** [comment]
same issue here - can't run any programs using DynamoRIO's drrun. wouldn't know if it's just a coincidence, but I'm also on a third-gen Ryzen

**Comment by trylab** [comment]
I also encountered this issue. Crash happened in function ``switch_modes_and_call``. I tried to debug it and hope the following logs will help.

The instructions in function ``switch_modes_and_call``:
```
dynamorio!switch_modes_and_call:
71f3ea1d 8bc4            mov     eax,esp
71f3ea1f 83c004          add     eax,4
71f3ea22 8b4c240c        mov     ecx,dword ptr [esp+0Ch]
71f3ea26 8b542410        mov     edx,dword ptr [esp+10h]
71f3ea2a 53              push    ebx
71f3ea2b 8b5c2418        mov     ebx,dword ptr [esp+18h]
71f3ea2f ea36eaf3713300  jmp     0033:71F3EA36
71f3ea36 41              inc     ecx
```

Once executed ``jmp     0033:71F3EA36``, crash will come shortly!
```
1:004> p
eax=005df600 ebx=180b28e0 ecx=00000138 edx=180b28e0 esi=180b193c edi=00000138
eip=71f3ea2b esp=005df5f8 ebp=005df6ac iopl=0         nv up ei pl nz ac pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000216
dynamorio!switch_modes_and_call+0xe:
71f3ea2b 8b5c2418        mov     ebx,dword ptr [esp+18h] ss:002b:005df610=00000000

1:004> t
eax=005df600 ebx=00000000 ecx=00000138 edx=180b28e0 esi=180b193c edi=00000138
eip=71f3ea2f esp=005df5f8 ebp=005df6ac iopl=0         nv up ei pl nz ac pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000216
dynamorio!switch_modes_and_call+0x12:
71f3ea2f ea36eaf3713300  jmp     0033:71F3EA36

1:004> t
eax=00000000 ebx=00000000 ecx=00000000 edx=00000000 esi=00000144 edi=00000080
eip=71f0fd96 esp=005df3c4 ebp=71f8ce7c iopl=0         nv up ei pl nz na po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000202
dynamorio!dynamorio_syscall_wow64_noedx+0x1f:
71f0fd96 8b0c24          mov     ecx,dword ptr [esp]  ss:002b:005df3c4=00002530

1:004> t
(1104.1cdc): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
eax=00000000 ebx=005df5d8 ecx=e2fbde54 edx=00000000 esi=180b193c edi=00000138
eip=71f3ea6b esp=005df5f8 ebp=005df6ac iopl=0         nv up ei pl nz na po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010202
dynamorio!switch_modes_and_call+0x4e:
71f3ea6b 5b              pop     ebx
```

Stack trace:
```
1:004> k
 # ChildEBP RetAddr  
00 005df6ac 71f16cf9 dynamorio!switch_modes_and_call+0x4e [C:\projects\dynamorio\build\build_release-32\core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s @ 1961] 
01 005df6dc 71f16b87 dynamorio!os_take_over_wow64_extra+0x49 [c:\projects\dynamorio\core\win32\os.c @ 2433] 
02 005dfbe8 71f1689f dynamorio!os_take_over_thread+0x197 [c:\projects\dynamorio\core\win32\os.c @ 2575] 
03 005dfc38 71e95746 dynamorio!os_take_over_all_unknown_threads+0x2cf [c:\projects\dynamorio\core\win32\os.c @ 2707] 
04 005dfc5c 71f0f81e dynamorio!dynamorio_take_over_threads+0xe6 [c:\projects\dynamorio\core\dynamo.c @ 2904] 
05 005dfc74 71f0fae8 dynamorio!auto_setup+0x3e [c:\projects\dynamorio\core\arch\x86_code.c @ 198] 
06 00000000 00000000 dynamorio!dynamo_auto_start+0x8 [C:\projects\dynamorio\build\build_release-32\core\CMakeFiles\dynamorio.dir\arch\x86\x86.asm.obj.s @ 1600] 
```

My CPU is **AMD R7 1700** and CPU virtualization is disabled.

**Comment by derekbruening** [comment]
It looks like you're using a 32-bit windbg, which will not be able to walk into the mode switch (the far jmp).  I would suggest using 64-bit windbg and walking across the mode switch to make sure everything is good there.  DR is assuming that 0x33 for a CS selector is what Windows has set up for a 64-bit code segment -- maybe that is not correct on your processor?  Though you would expect it to crash earlier in that case: but maybe the 32-bit windbg is confused about where the crash is.  If you look at a 64-bit process on your machine is CS 0x33?  What happens if you walk this same code with a 64-bit windbg: does it make the mode switch and run the 64-bit callee?  Where does it show the access violation?

**Comment by khang06** [comment]
getting the same thing as described here, win10 19041 w/ ryzen 3600

**Comment by trylab** [comment]
> It looks like you're using a 32-bit windbg, which will not be able to walk into the mode switch (the far jmp). I would suggest using 64-bit windbg and walking across the mode switch to make sure everything is good there. DR is assuming that 0x33 for a CS selector is what Windows has set up for a 64-bit code segment -- maybe that is not correct on your processor? Though you would expect it to crash earlier in that case: but maybe the 32-bit windbg is confused about where the crash is. If you look at a 64-bit process on your machine is CS 0x33? What happens if you walk this same code with a 64-bit windbg: does it make the mode switch and run the 64-bit callee? Where does it show the access violation?

It just works fine if I use 64-bit WinDbg to debug it. I mean if I steps into ``dynamorio!switch_modes_and_call`` instruction by instruction, the function works fine. But if I just let it run, it always crashes!

```
(e48.1cb8): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
dynamorio!switch_modes_and_call+0x4e:
74deea6b 5b              pop     ebx

1:004:x86> r
eax=00000000 ebx=00fceed0 ecx=a5edde54 edx=00000000 esi=1cf1193c edi=00000130
eip=74deea6b esp=00fceef0 ebp=00fcefa4 iopl=0         nv up ei pl nz ac pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010216
dynamorio!switch_modes_and_call+0x4e:
74deea6b 5b              pop     ebx

1:004:x86> k
 # ChildEBP RetAddr  
00 00fcefa4 74dc6cf9 dynamorio!switch_modes_and_call+0x4e [C:\projects\dynamorio\build\build_release-32\core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s @ 1961] 
01 00fcefd4 74dc6b87 dynamorio!os_take_over_wow64_extra+0x49 [c:\projects\dynamorio\core\win32\os.c @ 2433] 
02 00fcf4e0 74dc689f dynamorio!os_take_over_thread+0x197 [c:\projects\dynamorio\core\win32\os.c @ 2575] 
03 00fcf530 74d45746 dynamorio!os_take_over_all_unknown_threads+0x2cf [c:\projects\dynamorio\core\win32\os.c @ 2707] 
04 00fcf554 74dbf81e dynamorio!dynamorio_take_over_threads+0xe6 [c:\projects\dynamorio\core\dynamo.c @ 2904] 
05 00fcf56c 74dbfae8 dynamorio!auto_setup+0x3e [c:\projects\dynamorio\core\arch\x86_code.c @ 198] 
06 00000000 00000000 dynamorio!dynamo_auto_start+0x8 [C:\projects\dynamorio\build\build_release-32\core\CMakeFiles\dynamorio.dir\arch\x86\x86.asm.obj.s @ 1600] 
```

**Comment by derekbruening** [comment]
Hmm, so we're back to the original mystery of *why* there is an Access Violation at this `pop ebx` when the stack memory is accessible and `ss` looks normal; when we can presumably rule out the fault being on the *prior* instruction (with the current PC already having been advanced) b/c the prior is just "add esp,8"; and with weird exception info (not shown here but I assume this has the same `Parameter[0]: 00000003` as above)?

So there is some AMD-specific issue on the mode switch and the Windows kernel is raising it as an undocumented fault type?

But it seems timing-dependent based on the added delay loop making it less likely to happen, and stepping in the debugger making it go away?  But the target thread we're examining is suspended.  Does that mean one of the other threads is messing with it?

Could it be related to the function being called here: NtGetContextThread?  What if you change the code to keep the mode switch to remove the call to NtGetContextThread?  Then maybe we can isolate this to either the mode switch or to the query.  To make that code change: I guess remove the `call eax` from `switch_modes_and_call` (looks like esp will be restored properly), and maybe go zero out or do something with the context (maybe just fail: but there's an assert on that...hack it somehow).

Is the NtGetContextThread succeeding, in the debugger-step case *and* the crash case?

**Comment by tonybounty** [comment]
same problem, so I made a basic hello world x32/x64 in Visual Studio with default settings and using DynamoRIO 8.0.18417.

---
Host: Ubuntu Kernel 5.4.0, AMD Ryzen 7 2700
VM: VMWare Workstation (VT/RVI set, CPU perf. count set), Windows 10 1909, hello world **x32** :
CRASH
```
(3ec.176c): Break instruction exception - code 80000003 (first chance)
eax=00000000 ebx=00f8b000 ecx=eaca0000 edx=00000000 esi=014d2108 edi=77b6688c
eip=77c0eaa2 esp=0113efa0 ebp=0113efcc iopl=0         nv up ei pl zr na pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000246
ntdll!LdrpDoDebuggerBreak+0x2b:
77c0eaa2 cc              int     3
1:004> g
ModLoad: 6ddd0000 6df30000   C:\Users\tony\Desktop\DynamoRIO-Windows-8.0.18417\lib32\release\dynamorio.dll
(3ec.176c): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
eax=00000000 ebx=0113e898 ecx=8929de04 edx=00000000 esi=17a11d00 edi=00000054
eip=6de7f9db esp=0113e8b8 ebp=0113e96c iopl=0         nv up ei pl nz na pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010206
dynamorio!switch_modes_and_call+0x4e:
6de7f9db 5b              pop     ebx
1:004> .exr -1
ExceptionAddress: 6de7f9db (dynamorio!switch_modes_and_call+0x0000004e)
   ExceptionCode: c0000005 (Access violation)
  ExceptionFlags: 00000000
NumberParameters: 2
   Parameter[0]: 00000003
   Parameter[1]: 00000000
Attempt to write to address 00000000
```

---
Host: Ubuntu Kernel 5.4.0, AMD Ryzen 7 2700
VM: VMWare Workstation (VT/RVI set, CPU perf. count set), Windows 10 1909, hello world **x64** :
No CRASH 

---
Host: Laptop, Windows 10 1909, Intel i7 7500, hello world **x32** :
No CRASH

---
Host: Laptop, Windows 10 1909, Intel i7 7500, hello world **x64** :
No CRASH

**Comment by derekbruening** [comment]
Hoping someone with an affected Ryzen processor can help with some of the suggestions I put in the comment above: https://github.com/DynamoRIO/dynamorio/issues/4091#issuecomment-590552063

Another suggestion is to look at the WOW64 transition code in the system libraries, which the DR code is modeled after: is it using different code for Ryzen?

**Comment by electricworry** [comment]
I'm happy to give it a go. I've also got an Intel laptop to compare against so that might be helpful. As you know, I'm new to DynamoRIO and not comfortable with WOW64, so I might need a bit of guidance about what you would do and I can go away and try my best without taking up too much of your time.

Regarding analysing the crash as it happens within notepad.exe, I'm confused about how pop ebx/rbx would cause a failure unless the stack pointer's become invalid or intruded into an invalid address. So is analysis of that part of the problem/helpful? Is there a way for me to debug notepad before the crash so I can set a breakpoint on the failing function (switch_modes_and_call)?

And regarding analysis of the WOW64 transition code, do you mean that Windows might have different implementations depending on processor? Is there a particular dll I should be comparing between my two systems (I've got IDA Pro if it's of use) or is it more that a different code path might be taken and I need to follow that in the debugger? (And sorry, if it's the latter I might need a bit of guidance as to the method.)

**Comment by derekbruening** [comment]
> I'm confused about how pop ebx/rbx would cause a failure

Exactly, with valid `ss`, aligned and valid `esp`, executable `eip`.  It's looking like something weird: like a Ryzen processor bug on a `cs` change from 64-bit to 32-bit mode or something, surfacing as the strange exception with parameter 0==3.

The bug didn't show up previously on a single-step in the debugger (https://github.com/DynamoRIO/dynamorio/issues/4091#issuecomment-590223990) so it may not be easy to study inside windbg.

Some more ideas:
Look for a workaround, since it's always the `pop` that triggers it after the mode switch:
* Replace the `pop ebx` with an equivalent `mov ebx, dword ptr [esp]; add esp, 4`: does the bug go away?
* Add various barriers before the `pop ebx`: `mfence`, `lfence`, `sfence`, `cpuid`, `rdtsc`

More instructions for looking at the WOW64 mode transition code:
I would get a WOW64 process in 64-bit windbg and look at:
* `wow64cpu!X86SwitchTo64BitMode` which should have a far jmp like:
```
00000000`78b81910 ea2c3cb8783300  jmp     0033:78B83C2C
```
* `wow64cpu!CpuSimulate` which should have the return far jmp, sthg like:
```
00000000`78b83c29 41ff2e          jmp     fword ptr [r14] ds:00000000`0012edf0=00237d61ca01
```
Or at least that's where things were years ago.  Does anything look different?  Maybe there are just no `pop` instructions on the return path back through the 32-bit syscall wrappers and so natively the bug is not hit.

**Comment by byemypast** [comment]
I tried to copy the codes into MASM32 and compile it to see if the 'pop ebx' bug reappears, however it works successfully without any expcetion (on ryzen 1600x while crashed in drrun -- notepad).

Confused. It seems that it related to memory management malfunction(or some security feature?) in ryzen/windows.
Has anyone checked the function 'call eax' called in this exception? Maybe we can have a look at this.

Here's the MASM32 version code of switch_modes_and_load
`
.386
.model flat, stdcall  ;32 bit memory model
option casemap :none  ;case sensitive

include Test.inc

.data
strTitle db 'Helloworld!',0
strDone db 'Done!',0
.code

start:
	invoke MessageBox,0,addr strTitle,addr strTitle,0
	push 0
	push esp
	push 0
	push offset x64func
	call switch_modes_and_load
	pop eax 
	pop eax
	pop eax
	pop eax ; strange but it works.
	invoke MessageBox,0,addr strDone,addr strDone,0
	retn
	
switch_modes_and_load:

	;/* Routines to switch to 64-bit mode from 32-bit WOW64, make a 64-bit    
	;* call, and then return to 32-bit mode.    
	;*/    
	;/*    
	;* int switch_modes_and_load(void *ntdll64_LdrLoadDll,    
	;*                           UNICODE_STRING_64 *lib,    
	;*                           HANDLE *result)    
	;* XXX i#1633: this routine does not yet support ntdll64 > 4GB    
	;*/    
	mov eax, dword ptr [esp+4]
	mov ecx, dword ptr [esp+8]
	mov edx, dword ptr [esp+0ch]
	
	push ebx
	db 0eah
	dd offset sml_transfer_to_64
	db 033h, 0h

	
	
sml_transfer_to_64:
	;* Below here is executed in 64-bit mode, but with guarantees that    
	; no address is above 4GB, as this is a WOW64 process.    
	;    	
	;* Call LdrLoadDll to load 64-bit lib:    
	;*   LdrLoadDll(IN PWSTR DllPath OPTIONAL,    
	;*              IN PULONG DllCharacteristics OPTIONAL,    
	;*              IN PUNICODE_STRING DllName,    
	;*              OUT PVOID *DllHandle));    
       db 04ch,08bh,0cah ;mov r9, rdx : 4th arg: result     
       db 04ch,08bh,0c1h ;mov r8, rcx : 3rd arg: lib */    
       push     0        ;       /* slot for &DllCharacteristics */    
       lea      edx, dword ptr [esp] ;/* 2nd arg: &DllCharacteristics */    
       xor      ecx, ecx ;       /* 1st arg: DllPath = NULL */    
       ;/* save WOW64 state */    
       db 041h 
       push     esp ;/* push r12 */    
       db 041h 
       push     ebp ;/* push r13 */
       db 041h    
       push     esi ;/* push r14 */    
       db 041h 
       push     edi ;/* push r15 */    
       ;/* align the stack pointer */    
       mov      ebx, esp        ;/* save esp in callee-preserved reg */    
       sub      esp, 32         ;/* call conv */    
       and      esp, 0fffffff0h ;/* align to 16-byte boundary */    
       call     eax    
       mov      esp, ebx        ;/* restore esp */    
       ;/* restore WOW64 state */    
       db 041h 
       pop      edi ;/* pop r15 */    
       db 041h 
       pop      esi ;/* pop r14 */    
       db 041h 
       pop      ebp ;/* pop r13 */    
       db 041h 
       pop      esp ;/* pop r12 */    
       ;/* far jmp to next instr w/ 32-bit switch: jmp 0023:<sml_return_to_32> */    
       push     offset sml_return_to_32  ;/* 8-byte push */    
       mov      dword ptr [esp + 4], 023h ;/* top 4 bytes of prev push */    
       jmp      fword ptr [esp]    
sml_return_to_32:    
       add      esp, 16         ;/* clean up far jmp target and &DllCharacteristics */    
       pop      ebx             ;/* restore callee-saved reg */    
       ret                      ;/* return value already in eax */

x64func:
	db 049h, 08bh, 0d7h ;mov rdx,r15
	retn       
end start


`

**Comment by electricworry** [comment]
Thanks @derekbruening, I'll give it a go. I won't get a chance until Sunday or Monday at the earliest due to other work. Of course, @byemypast might be far ahead of me by that point.

**Comment by derekbruening** [comment]
> Has anyone checked the function 'call eax'

I suggested in https://github.com/DynamoRIO/dynamorio/issues/4091#issuecomment-590552063 removing the call to see if it made a difference.  The call, at least in the original case above, was a call to NtGetContextThread.  See the linked comment where I also asked whether that call returned 0 or not in the crashing case.

**Comment by electricworry** [comment]
Hello, I've done some things, largely corroborating what others have done already. Like original poster observed, sometimes notepad will run ok about 1/100.

> Replace the pop ebx with an equivalent mov ebx, dword ptr [esp]; add esp, 4: does the bug go away?

No, the mov ebx, dword ptr[esp] fails.

> Add various barriers before the pop ebx: mfence, lfence, sfence, cpuid, rdtsc

This doesn't help either.

> I guess remove the call eax from switch_modes_and_call

When I remove the "call eax" from switch_modes_and_call the problem does seem to go away. I ran it about 20 times and not one crash, so very repeatable avoidance of the issue.

I can also confirm that introducing a loop before pop ebx alleviates the problem, but I agree this isn't an adequate solution.

I decided to try the int 3 breakpoint before the "pop ebx" so I ran "bin32\drrun.exe -debug -msgbox_mask 15 -- notepad" and continued through and indeed the problem went away. It doesn't really matter where the int 3 is (just before the pop, or further up above the call eax); I can just hit 'g' to continue and the program doesn't crash. Indeed, without the int 3, if I just set up an unresolved breakpoint on dynamorio!switch_modes_and_call and go on each occurence then the problem is gone too.

> I also asked whether that call returned 0 or not in the crashing case

That would be the value 0x0 in eax, correct? So, yes the call eax returned zero.

Now that I'm up to speed, what's the best next step? It sounds another thread is suspected. Is the thinking that the other thread might be altering the memory protections during the execution of switch_modes_and_call, and that by halting program with the debugger (or introducing the loop) we're allowing that other thread to be just ahead of the game? This seems to suggest that we can't trust what's observable in the debugger.

Would it be helpful to modify switch_modes_and_call so that it dumps some relevant information to a dialog just before the pop? If so, what sort of information should we gather? (Just throwing that idea out there.)

As for the other suggestion to "look at the WOW64 transition code in the system libraries", I could give that a go. Do you know what library I need to examine?

Hopefully that's covered all of the suggestions so far; sorry if I've missed something important.

**Comment by derekbruening** [comment]
> > Replace the pop ebx with an equivalent mov ebx, dword ptr [esp]; add esp, 4: does the bug go away?
> 
> No, the mov ebx, dword ptr[esp] fails.

With the same weird `ExceptionCode=c0000005, Parameter[0]=00000003`?

What if we avoid the `ss:` default segment prefix by not having `esp` as the base: sthg like "mov ebx, esp; mov ebx, dword ptr [ebx]"?

> Now that I'm up to speed, what's the best next step? It sounds another thread is suspected. Is the thinking that the other thread might be altering the memory protections during the execution of switch_modes_and_call, and that by halting program with the debugger (or introducing the loop) we're allowing that other thread to be just ahead of the game? This seems to suggest that we can't trust what's observable in the debugger.

A couple of thoughts here:
+ The thread whose context we're querying should be suspended: we just suspended it so we can safely acquire the context to take it over.  Does windbg confirm it's suspended (`~` shows the suspend count)?
+ How many other threads are there besides the DR takeover thread and this target thread?
  - Have we already taken some of those over?
  - If not, could we change the takeover order?
  - Is it always the same thread whose context querying triggers the takeover thread to crash?
  - What are these other threads doing during this time?
  - What if we suspend *all* of the other threads all at once, and then go back through to acquire contexts and take over, instead of the current one-at-a-time suspend-and-take-over approach?
+ The undocumented Parameter[0]=00000003 indicates this is not a vanilla access violation, so I would expect something more complex/weird than page protection races (plus it would be odd for another thread to change the private stack protections of a thread).  Maybe we should ask some Windows kernel folks about that 3 value.

> As for the other suggestion to "look at the WOW64 transition code in the system libraries", I could give that a go. Do you know what library I need to examine?

I put that information in https://github.com/DynamoRIO/dynamorio/issues/4091#issuecomment-665932712

This all seems very strange: I would expect a stack reference issue after a mode switch to be some bug particular to one hardware thread.  Could it be related to a thread context switch and the processor+OS not properly swapping/restoring some value?  Xref a known AMD processor bug of that nature that we hit in #3356 where the Linux kernel and DR have to take extra steps to work around the hardware bug.  Maybe if we get a thread context switch in the wrong spot things get messed up, and inserting the loop changes which software thread is on which hardware thread.

**Comment by derekbruening** [comment]
> Would it be helpful to modify switch_modes_and_call so that it dumps some relevant information to a dialog just before the pop? If so, what sort of information should we gather? (Just throwing that idea out there.)

If the avoid-ss attempt works, I would record `ss`: but we may need not just the selector but also the hidden parts and the LDT/GDT entries.  We could start with the selector at least.

**Comment by electricworry** [comment]
> > > Replace the pop ebx with an equivalent mov ebx, dword ptr [esp]; add esp, 4: does the bug go away?
> > 
> > 
> > No, the mov ebx, dword ptr[esp] fails.
> 
> With the same weird `ExceptionCode=c0000005, Parameter[0]=00000003`?

Yes. Sorry, I should have said. Same ExceptionCode=c0000005, Parameter[0]=00000003, and on our new mov ebx, dword ptr [esp] instruction.

> What if we avoid the `ss:` default segment prefix by not having `esp` as the base: sthg like "mov ebx, esp; mov ebx, dword ptr [ebx]"?

Interesting. It still has the same exception, but it's not on the replacement instructions, it's the ret that fails:

```
0:003> g
(10f0.235c): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
dynamorio!switch_modes_and_call+0x57:
7a3578f4 c3              ret
0:000:x86> .exr -1
ExceptionAddress: 7a3578f4 (dynamorio!switch_modes_and_call+0x00000057)
   ExceptionCode: c0000005 (Access violation)
  ExceptionFlags: 00000000
NumberParameters: 2
   Parameter[0]: 00000003
   Parameter[1]: 00000000
Attempt to write to address 00000000
```
Here's the asm in windbg just to confirm correctness:
```
7a3578ea 83c408          add     esp,8
7a3578ed 8bdc            mov     ebx,esp
7a3578ef 8b1b            mov     ebx,dword ptr [ebx]
7a3578f1 83c404          add     esp,4
7a3578f4 c3              ret
dynamorio!load_dynamo:
```
> If the avoid-ss attempt works, I would record ss: but we may need not just the selector but also the hidden parts and the LDT/GDT entries. We could start with the selector at least.

Since I've got the DynamoRIO Notice popping up, I just threw in a "mov eax, ss" before the ret, and the output showed eax=0x0000002b. ss=002b should be right/valid I believe.

That's all I had time to look at so far. I'll continue with the rest of the suggestions tomorrow, but if you know some Windows kernel people that would be great. I wouldn't know where to begin.

**Comment by derekbruening** [comment]
> Interesting. It still has the same exception, but it's not on the replacement instructions, it's the ret that fails:

Aha!  That lends credence to the theory that `ss` is messed up.  We could replace the `ret` with an alternative that avoids an explicit `esp` base as well and it would then not crash there either, but back in the caller, presumably.

I wonder if this `Parameter[0]=00000003` can be reproduced in a standalone app with a messed-up `ss`.

But if the visible selector seems correct, I guess we have to dig deeper to figure out which part of it is wrong and how to fix it.

What if a yield is used instead of the loop?  I remember the loop avoided the crash only *most* of the time (up above the iters had to be increased): does a yield avoid it every time?  I.e., is it that the loop triggers a context switch for *this* takeover thread, and not that it is waiting for some other thread to do something in parallel.

**Comment by electricworry** [comment]
Forgive my ignorance but what's a yield instruction in Intel/AMD64? I can see there's such a thing in ARM but can't find anything equivalent.

**Comment by derekbruening** [comment]
> Forgive my ignorance but what's a yield instruction in Intel/AMD64? I can see there's such a thing in ARM but can't find anything equivalent.

System call NtYieldExecution: DR has a wrapper os_thread_yield().

**Comment by electricworry** [comment]
Sorry I might need more help. I'm not very fluent with Windows and Visual Studio.

I've added "DECL_EXTERN(os_thread_yield)" and put in a "call os_thread_yield", but I get linking failures in various projects, e.g.:

```
Creating library C:/Users/Worry/projects/dynamorio/build/lib32/drinjectlib.lib and object C:/Users/Worry/projects/dynamorio/build/lib32/drinjectlib.exp
drlibc.lib(drlibc_x86.asm.obj) : error LNK2019: unresolved external symbol _os_thread_yield referenced in function _switch_modes_and_call [C:\Users\Worry\projects\dynamorio\build\core\drinjectlib.vcxproj]
C:\Users\Worry\projects\dynamorio\build\lib32\drinjectlib.dll : fatal error LNK1120: 1 unresolved externals [C:\Users\Worry\projects\dynamorio\build\core\drinjectlib.vcxproj]
```
I need to add the COFF file os.obj to various failing projects? I'm not sure of the VS solution method of doing that.

**Comment by derekbruening** [comment]
Use `nt_yield` instead?

**Comment by electricworry** [comment]
No, I'm sorry I get similar linking issues with that. And I've tried NtYieldExecution too. I think I just need to figure out the right change to core/CMakeLists.txt so that the function (or one of them) can be resolved.

The implementations of the various functions appear to exist in the following places:

- os_thread_yield
  - dynamorio_static.lib
  - os.obj
- nt_yield
  - dynamorio_static.lib
  - ntdll.obj
- NtYieldExecution
  - ntdll_imports.dll
  - ntdll_imports.obj

**Comment by derekbruening** [comment]
Just ignore drinjectlib and rebuild only dynamorio.dll.  No need to get the yield into drinjectlib.

**Comment by electricworry** [comment]
Right you are.. I didn't think to check if what was building was enough, sorry.

It was a great idea, but sadly no improvement; the call to os_thread_yield fails. This whole thing is really weird. This whole bug is way beyond my CPU/kernel knowledge:

```
0:003> g
(2360.934): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
dynamorio!switch_modes_and_call+0x4e:
516478eb e8c0c0f7ff      call    dynamorio!os_thread_yield (515c39b0)
0:000:x86> .exr -1
ExceptionAddress: 516478eb (dynamorio!switch_modes_and_call+0x0000004e)
   ExceptionCode: c0000005 (Access violation)
  ExceptionFlags: 00000000
NumberParameters: 2
   Parameter[0]: 00000003
   Parameter[1]: 00000000
Attempt to write to address 00000000
0:000:x86> r
eax=00000000 ebx=022eea88 ecx=2a16cbc4 edx=00000000 esi=1d1a9f00 edi=022eeb38
eip=516478eb esp=022eeaa8 ebp=022eeadc iopl=0         nv up ei pl nz na po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010202
dynamorio!switch_modes_and_call+0x4e:
516478eb e8c0c0f7ff      call    dynamorio!os_thread_yield (515c39b0)
```

**Comment by derekbruening** [comment]
Ah yeah, the call uses the stack, so it's hitting the same crash as the pop, the mov, and the ret.  We'd have to inline the syscall: which for WOW64 is very non-trivial: you have to switch modes back to 64-bit.  What if we inlined it while still in 64-bit mode?  That's much easier: move the immediate to eax (disassemble the wrapper for your Windows version to see the value to use) and execute `syscall` (making sure the 32-bit compiler will encode that properly).

For exploring the `ss` segment more: try the `dg` windbg command at an earlier normal point and at the crash point.

Another step could be to make a standalone reproducer that does not involve DR but just two threads, one running the mode switch and NtGetContextThread call after suspending the other, which will help zero in on what the trigger is, and will help pointing other people at it for help.

**Comment by electricworry** [comment]
Small update. Not made much progress today as my boss was helping me get up to speed with WOW64. So tomorrow I should be in a better place to look at the transition code.

**Comment by electricworry** [comment]
I've got some news. Please remember I'm new to Windows internals so apologies if some of my explanations are painfully obvious. :)

First I spent some time looking at the Windows mode switch code. Only at a basic level; like you said, it is complicated. However, when looking at the method and the types of instructions I could see nothing wrong with the DynamoRIO switch code. Nothing that would indicate that one shouldn't use pops. The jumps all looked sane, as you would expect (otherwise it should crash on Intel too, right?).

Earlier discussion made the hypothesis that one of the other threads might be exacerbating the issue. I thought it would be interesting to at least prove or disprove that. So I built a basic hello world application and tested with that. I would use the msgbox_mask switch to get in early, and then if I let it run it would crash as per notepad.exe. At the point of attaching with the debugger here was the stack:

```
01 0019e99c 7935788b dynamorio!nt_messagebox+0x130 [c:\users\worry\projects\dynamorio\core\win32\ntdll.c @ 3691]
02 0019e9c4 79190c4c dynamorio!debugbox+0x4b [c:\users\worry\projects\dynamorio\core\win32\os.c @ 5291]
03 0019f1dc 790cc96c dynamorio!d_r_notify+0x1dc [c:\users\worry\projects\dynamorio\core\utils.c @ 1933]
04 0019fabc 79338b51 dynamorio!dynamorio_app_init+0x22c [c:\users\worry\projects\dynamorio\core\dynamo.c @ 513]
05 0019fb24 793d7108 dynamorio!auto_setup+0x21 [c:\users\worry\projects\dynamorio\core\arch\x86_code.c @ 169]
06 00000000 00000000 dynamorio!dynamo_auto_start+0x8 [C:\Users\Worry\projects\dynamorio\build\core\x86.asm_core.s @ 1600]
```

Looks good. But I wondered why there were so many threads at this point. WinDbg showed 4. This bothered me because my app is definitely single threaded. Since I'm new to DR I wondered if the threads were started somewhere in the DR chain as part of the injection, so I broke in earlier using the "#ifdef LOAD_DYNAMO_DEBUGBREAK" which places a very large loop in load_dynamo(). This should be before any other threads can be created. Still 4 threads.

So queries about what all these threads are brought me first to [this](https://stackoverflow.com/questions/34822072/why-does-windows-10-start-extra-threads-in-my-program), then to [this](https://blogs.blackberry.com/en/2017/10/windows-10-parallel-loading-breakdown).

Setting the MaxLoaderThreads to 1 for my application hello.exe reduced the total number of threads to 2. And now the problem is gone. I can DRrun the application (with or without msgbox_mask and a debugger) 100% of the time. The same applies for notepad with similar registry changes.

So I don't have a why yet, but this would appear to be a workaround for now. And I think it vindicates the DR mode switch code.

Does anyone have any suspicions? It looks to me like Windows 10 parallel loading is trashing something? Could there be an AMD specific bug in the parallel loading code?

**Comment by electricworry** [comment]
I'm thinking it might be a good idea for me to try modifying the PEB to set the loader threads to 1. Apparently that's can be done. Any examples of modifying the PEB from the injector side before ResumeThread?

**Comment by tonybounty** [comment]
I confirm that when I set `MaxLoaderThreads` to 1 in `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\TestDynamo.exe`, I no longer have issues with drrun and WOW64 app on my Ryzen (Win10 x64 VM).

**Comment by derekbruening** [comment]
Thank you for looking further into this!

> Since I'm new to DR I wondered if the threads were started somewhere in the DR chain as part of the injection,

Plain DR (i.e., with no client) never creates new threads: it considers it non-transparent to muck with the threads and a risk to mess up the app.  (Clients can create threads at their own risk; client-or-user-initiated "nudges" use temporary threads.)

> Setting the MaxLoaderThreads to 1 for my application hello.exe reduced the total number of threads to 2. 

According to the blog you pointed at, setting MaxLoaderThreads to 1 should create 0 thread pool threads: so why is there still a 2nd thread, and why doesn't it cause a problem, since DR will run the same mode switching code to take it over?  Is it not a Tpp thread and it's created later?  Sometimes Windows creates temporary threads for things like CTRL_SHUTDOWN.

The problem with limiting the thread pool is that it's going to affect the app which might be relying on having the pool there (right?) so it does not seem an ideal permanent solution, but it's good to have a workaround.

If we used "earliest injection" (taking over at the very first user mode instruction), the problem would go away for launch-the-app usage b/c we wouldn't need to take over other threads at DR init time.  However, while we have code to do this, it is very difficult to support clients at that early point, since clients want to use libraries.

Just confirming to myself here: DR needs to suspend and get the context of threads for other reasons than attaching with >1 thread, so presumably this problem can occur at later points of execution.  E.g., if the app suspends a thread, DR has to intervene; if a tool wants to suspend threads (e.g., Dr. Memory suspends the world to perform its leak scan), DR will need to get the thread context.  But it looks like a 32-bit NtGetContextThread is sufficient for those cases: it is only for the attach case that thread_get_context_64() with this mode switch is needed.  So this is limited to attach time only.

**Comment by derekbruening** [comment]
> I'm thinking it might be a good idea for me to try modifying the PEB to set the loader threads to 1. Apparently that's can be done. Any examples of modifying the PEB from the injector side before ResumeThread?

DR has to routinely read and mess with the PEB (even creating a private copy of it and swapping back and forth) so there is plenty of existing code doing that: see get_peb(), get_own_peb(), etc.  You may need to update the RTL_USER_PROCESS_PARAMETERS struct def.

**Comment by derekbruening** [comment]
Another idea: See if re-instating the ss selector fixes the segment info: add instructions to move 0x2b into `ss` right after the mode switch and before any `esp` reference.

**Comment by hunterbr72** [comment]
As mentioned in the user group, same bug here. We run it with a debug build of DynRio now. Interesting is that it doesn't always crash. HTH, tbh I don't know enough about the DynRio internals to track that any further, pls advise if we should test anything else ? 

Machine:
Processor AMD Ryzen Threadripper 2990WX
Microsoft Windows 10 Pro Version 10.0.19041 Build 19041

testsample0_32_dyn.exe = very simple single threaded app.

Cmd:
------
C:\tools\dev>C:\tools\dev\git\dynamorioX32_dbg\build\bin32\drrun.exe -verbose -debug -msgbox_mask 15 -- "C:\tools\DDR\samples\testsample0_32_dyn.exe"
INFO: targeting application: "C:\tools\DDR\samples\testsample0_32_dyn.exe"
INFO: app cmdline:  "C:\tools\DDR\samples\testsample0_32_dyn.exe"
INFO: configuration directory is "C:\Users\Dex Dexter/dynamorio"
INFO: created child with pid 1280 for C:\tools\DDR\samples\testsample0_32_dyn.exe
INFO: waiting forever for app to exit...
<Starting application C:\tools\DDR\samples\testsample0_32_dyn.exe (1280)>
<Running on newer-than-this-build "Microsoft Windows 10-2004 x64">
<Early threads found>
<Initial options = -no_dynamic_options -code_api -probe_api -msgbox_mask 15 -stack_size 56K -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct -no_aslr_dr >
<Application C:\tools\DDR\samples\testsample0_32_dyn.exe (1280).  DynamoRIO internal crash at PC 0x7551965b.  Please report this at http://dynamorio.org/issues/.  Program aborted.
0xc0000005 0x00000000 0x7551965b 0x7551965b 0x00000003 0x00000000
Base: 0x75200000
Registers: eax=0x00000000 ebx=0x00bcebf0 ecx=0x42acdbf4 edx=0x00000000
        esi=0x1653e340 edi=0x00bceca0 esp=0x00bcec10 ebp=0x00bcec44
        eflags=0x0001
version 8.0.33333, custom build
-no_dynamic_options -code_api -probe_api -msgbox_mask 15 -stack_size 56K -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct -no_aslr_dr
0x00bcec44 0x754a1a19
0x00bcec70 0x754a1ea8
0x00bcf7d0 0x75494978
0x00bcf810 0x75210498
0x00bcf844 0x7547aa72
0x00bcf8b0 0x75518e88>



Windbg:
----------
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\atlmfc.natvis'
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\concurrency.natvis'
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\cpp_rest.natvis'
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\stl.natvis'
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\Windows.Data.Json.natvis'
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\Windows.Devices.Geolocation.natvis'
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\Windows.Devices.Sensors.natvis'
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\Windows.Media.natvis'
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\windows.natvis'
NatVis script unloaded from 'C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\Visualizers\winrt.natvis'

*** wait with pending attach


************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
OK                                             C:\Users\Dex Dexter\Documents\GitHub\DynDataResolver\ddr_client\Debug
OK                                             C:\Windows/system32
OK                                             C:\tools\DDR\DynamoRIO-Windows-8.0.0-1\ext\lib32\release
OK                                             C:\tools\DDR\samples
OK                                             C:\tools\DDR\DynamoRIO-Windows-8.0.0-1\lib32\release
Symbol search path is: srv*;C:\Users\Dex Dexter\Documents\GitHub\DynDataResolver\ddr_client\Debug;C:\Windows/system32;C:\tools\DDR\DynamoRIO-Windows-8.0.0-1\ext\lib32\release;C:\tools\DDR\samples;C:\tools\DDR\DynamoRIO-Windows-8.0.0-1\lib32\release
Executable search path is:
ModLoad: 00000000`004b0000 00000000`004b7000   C:\tools\DDR\samples\testsample0_32_dyn.exe
ModLoad: 00007ffb`42a30000 00007ffb`42c25000   C:\Windows\SYSTEM32\ntdll.dll
ModLoad: 00000000`779b0000 00000000`77b53000   ntdll.dll
ModLoad: 00007ffb`40f90000 00007ffb`40fe9000   C:\Windows\System32\wow64.dll
ModLoad: 00007ffb`42120000 00007ffb`421a3000   C:\Windows\System32\wow64win.dll
ModLoad: 00000000`779a0000 00000000`779aa000   C:\Windows\System32\wow64cpu.dll
ModLoad: 00000000`76770000 00000000`76860000   KERNEL32.dll
ModLoad: 00000000`760c0000 00000000`762d3000   KERNELBASE.dll
ModLoad: 00000000`77150000 00000000`77270000   ucrtbase.dll
ModLoad: 00000000`756c0000 00000000`756d4000   VCRUNTIME140.dll
ModLoad: 00000000`75200000 00000000`756b9000   dynamorio.dll
(500.1258): Break instruction exception - code 80000003 (first chance)
ntdll!DbgBreakPointProcessing initial command '$><C:\tools\dev\git\dynamorio\tools\windbg-scripts\load_symsWOW64'
:
00007ffb`42acf8b0 cc              int     3

0:003> $><C:\tools\dev\git\dynamorio\tools\windbg-scripts\load_symsWOW64

Memory access error at ') + @$t0+5) & 0xfffff000);;$$ Check magic values to avoid executing random command w/o DynamoRIO;.if (dwo(@$t1) = b1d2ae58) {;    .if (dwo(@$t1 + 4) = ca50c356) {;        .if (dwo(@$t1 + 8) = 63000089) {;            .if (dwo(@$t1 + c) = 3fa898f0) {;                aS /c ${/v:loadpriv} .printf "%ma", @$t1 + 2c;                .block { ${loadpriv} };                ad ${/v:loadpriv};            } .else {;                .echo "DynamoRIO not detected";            };        } .else {;            .echo "DynamoRIO not detected";        };    } .else {;        .echo "DynamoRIO not detected";    };} .else {;    .echo "DynamoRIO not detected";};'

0:003> g

(500.568): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.

dynamorio!dynamorio_earliest_init_takeover+0x119:
7551965b 5b              pop     ebx
0:000:x86> r

eax=00000000 ebx=00bcebf0 ecx=42acdbf4 edx=00000000 esi=1653e340 edi=00bceca0
eip=7551965b esp=00bcec10 ebp=00bcec44 iopl=0         nv up ei pl nz ac po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010212

dynamorio!dynamorio_earliest_init_takeover+0x119:

7551965b 5b              pop     ebx
0:000:x86> kn

 # ChildEBP RetAddr

WARNING: Stack unwind information not available. Following frames may be wrong.
00 00bcec44 754a1a19 dynamorio!dynamorio_earliest_init_takeover+0x119
01 00bcec70 754a1ea8 dynamorio!dr_mcontext_to_context+0x2c49
02 00bcf7d0 75494978 dynamorio!dr_mcontext_to_context+0x30d8
03 00bcf810 75210498 dynamorio!dr_insert_restore_fpstate+0x31808
04 00bcf844 7547aa72 dynamorio!dynamorio_app_init+0x35a8
05 00bcf8b0 75518e88 dynamorio!dr_insert_restore_fpstate+0x17902
06 00000000 00000000 dynamorio!dynamo_auto_start+0x8
0:000:x86> g

ntdll!NtWaitForWorkViaWorkerFactory+0x14:

00007ffb`42acf814 c3              ret

**Comment by cha5126568** [comment]
Setting MaxLoaderThreads to 1 seems to solve the problem, but it doesn't seem to be. I am trying to use dynaomorio to fuzz my target with winafl (by modifying the winafl module to support manual mapping.)

The solution to setting MaxLoaderThreads to 1 does not cause a crash, but "sometimes" dynaomorio does not feedback the map information to the fuzzer. (See the picture below.)
![](https://kong.re.kr/img/diff.png)
Therefore, there is an issue where the stability value continues to decline, which is preventing my fuzzing.

Is this different from the threading issue mentioned above?

Anyone else have experienced this problem?

**Comment by derekbruening** [comment]
@hunterbr72 -- Thank you for the post (I added markdown triple backtics to verbatim-quote for easier reading).  So windbg sees `ss` as having the exact same properties at the crash.  I guess I don't see that as invalidating the theory of `ss` being messed up b/c it could be something internal to the processor, though it would be nicer if windbg would just show us some messed-up visible state to explain the access violation.

@cha5126568 -- Could you elaborate on what is causing this map info to not be present, if it's not a crash?  Is it some other crash later during the run?

**Comment by norbertmm** [comment]
I confirm the Bug for AMD Ryzen Threadripper 3990X on Windows 10 Enterprise, 1909, and would like help testing workarounds and/or possible fixes for the issue. I am immediately ready to compile and test modifications to DynamoRIO and/or DrMemory.

**Comment by derekbruening** [comment]
@norbertmm -- Thank you for the offer.  I wonder if you could try the suggestion above at https://github.com/DynamoRIO/dynamorio/issues/4091#issuecomment-703715677 by modifying `core/drlibc/drlibc_x86.asm`:

```
smc_return_to_32:
        add      esp, 8          /* clean up far jmp target */
        pop      ebx             /* restore callee-saved reg */
        ret                      /* return value already in eax */
```
becomes:
```
smc_return_to_32:
        add      esp, 8          /* clean up far jmp target */
        mov      eax, HEX(2b)
        mov      ss, eax
        pop      ebx             /* restore callee-saved reg */
        ret                      /* return value already in eax */
```

Don't really have high hopes but seems worth a shot since it's simple.

IMHO the best solution may be the one mentioned in https://github.com/DynamoRIO/dynamorio/issues/4091#issuecomment-668072660:

+ Change the takeover code to suspend all of the other threads simultaneously, and only then march through them and invoke the mode-switched NtGetContextThread, to eliminate whatever the other threads are doing to trigger the problem.

I wonder if there is a way to test this out to see if it would work: have windbg suspend the other threads at the top of `os_take_over_all_unknown_threads()` and resume each when it's targeted in `os_take_over_thread()` (or maybe that's not necessary: double-suspend is ok).

Or, just implement this: I think it should be simple.  Copy the loop in `os_take_over_all_unknown_threads()` where `os_take_over_thread()` is called and have the first copy invoke `nt_thread_suspend()` instead of `os_take_over_thread`, and have the 2nd loop pass `true` for `suspended`.

**Comment by norbertmm** [comment]
It works!!!  
```mov  ebx, HEX(2b)
mov  ss, ebx```
does the job. I will analyze futher and provide more details soon.

**Comment by norbertmm** [comment]
![DebugSession1](https://user-images.githubusercontent.com/56049856/95672822-12082d80-0ba4-11eb-83fd-03fd0b6a3db3.png)

**Comment by cha5126568** [comment]
Sorry, the problem I had was a target-related race condition issue.

My target doesn't internally load a file that has been loaded once, if it has been loaded once using the hash map (hashmap is saved as a file.)

So, to solve this, I wrote a code that deletes the file on every run, but I don't understand why, but it worked fine on a my laptop computer with Intel CPU, but not on RyZen bare metal.
(But I really don't know why, I'll analyze it in detail when I have time.)

So I thought it was related to this issue, but when I changed the part of deleting the file in the direction of leaving the file open without FILE_SHARE_WRITE, it resolved.




I tested also https://github.com/DynamoRIO/dynamorio/issues/4091#issuecomment-703715677 and it works properly!

**Comment by derekbruening** [comment]
That's awesome that we've found a fix.  Thank you for testing it.

For the official patch we should probably dynamically determine the `0x2b` selector value rather than hardcoding it, maybe putting it in a variable at init time that is then loaded by the assembly.

**Comment by derekbruening** [comment]
It would be great if someone who has an AMD processor hitting this problem could test the official proposed fix in PR #4485 as a sanity check to ensure it replicates the smaller patch here.

**Comment by hunterbr72** [comment]
> 
> 
> It would be great if someone who has an AMD processor hitting this problem could test the official proposed fix in PR #4485 as a sanity check to ensure it replicates the smaller patch here.

Thx a lot Derek, I have built the patched DynRio version and provided it to the user who is using my client dll and asked him to test the patch on his AMD box. I ll post here as soon as I have an answer.

**Comment by hunterbr72** [comment]
P.S. it is running fine on my Intel box too... just for the protocol.

**Comment by norbertmm** [comment]
@derekbruening
Your fix does use EAX for SS value, overwriting the function call's return value. Thats probably not intented.
I suggest using EBX instead, as it will be restored immediately after anyways.

**Comment by hunterbr72** [comment]
I confirm that the new version doesn't crash anymore and works on the VMware VM below:

Processor AMD Ryzen Threadripper 2990WX
Microsoft Windows 10 Pro Version 10.0.19041 Build 19041

Thx a lot to Derek and everyone working on it for the fast fix !

**Comment by electricworry** [comment]
Sorry I've not been able to look at this for the last couple of months. Just been too busy with work and to be honest, when you can't trust the CPU because of a bug I don't really know what I'm doing.

I too have tested the fix in i4091-amd-ss-mode-switch branch and can confirm that it doesn't crash any more. However I share @norbertmm's concern that EAX is being overwritten. Kudos to @derekbruening for finding a suitable fix!

Can anyone please explain what's going on here? Is it not always possible to trust the value read from registers like ss? i.e. is there internal data behind it (which is not readable) which could be corrupt and will only get corrected on a subsequent overwrite? Also, is this something that should be raised with AMD?

Cheers.

**Comment by norbertmm** [comment]
@electricworry See section 3.4.2 and 3.4.3 of [Intel® 64 and IA-32 ArchitecturesSoftware Developer’s Manual](https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3a-part-1-manual.pdf) for a short explanation about the visible and hidden parts of the _segment registers_.

Basically it states: If you write a _segment selector_ value to a segment register, the CPU will load an entry from the indexed LDT/GDT table which will be cached and used subsequently. But you cannot inspect or read it back.

The _write to SS bugfix_, which is proposed, triggers a re-read of the GDT descriptor by the CPU which fixes a (partly?) incorrect entry in the CPUs cache (register).
It's still unclear how the CPU's internal cache gets messed up: Due to a CPU bug or due to some unexpected actions of the OS DLL loader threads in conjunction with the 64-bit NtGetContextThread() function call in a WOW64 process.

At least this is what I have learned so far.

**Comment by electricworry** [comment]
Cheers for that explanation, @norbertmm.

**Comment by derekbruening** [comment]
> @derekbruening
> Your fix does use EAX for SS value, overwriting the function call's return value. That is probably not intented.
> I suggest using EBX instead, as it will be restored immediately after anyways.

+1, thank you.  Switched to ebx.

**Comment by electricworry** [comment]
Thanks! I have retested and still confirm that the crash issue is resolved.

**Comment by derekbruening** [comment]
I triggered a new build so there's a Windows binary at https://github.com/DynamoRIO/dynamorio/releases/tag/cronbuild-8.0.18547 with the fix.

**Comment by norbertmm** [comment]
Some debugging with WinDbg 64 revealed this code on return path from 64bit system calls to 32bit application.
So **Windows always reloads the SS register with 0x2B.**
```
wow64cpu!RunSimulatedCode+0x37:
00000000`76ea1657 418b7d20        mov     edi,dword ptr [r13+20h]
00000000`76ea165b 418b7524        mov     esi,dword ptr [r13+24h]
00000000`76ea165f 418b5d28        mov     ebx,dword ptr [r13+28h]
00000000`76ea1663 418b6d38        mov     ebp,dword ptr [r13+38h]
00000000`76ea1667 418b4534        mov     eax,dword ptr [r13+34h]
00000000`76ea166b 4c8bf4          mov     r14,rsp
00000000`76ea166e c744240423000000 mov     dword ptr [rsp+4],23h
00000000`76ea1676 41b82b000000    mov     r8d,2Bh
00000000`76ea167c 418ed0          mov     ss,r8w
00000000`76ea167f 458b4d3c        mov     r9d,dword ptr [r13+3Ch]
00000000`76ea1683 44890c24        mov     dword ptr [rsp],r9d
00000000`76ea1687 418b6548        mov     esp,dword ptr [r13+48h]
00000000`76ea168b 41ff2e          jmp     fword ptr [r14]
```
Although, in 64bit code SS is set to 2B also:
```
Breakpoint 1 hit
wow64!RunCpuSimulation:
00007ffa`4cfac770 4883ec28        sub     rsp,28h
0:000> k
 # Child-SP          RetAddr           Call Site
00 00000000`0086df38 00007ffa`4cfb856d wow64!RunCpuSimulation
01 00000000`0086df40 00007ffa`4d7d9054 wow64!Wow64KiUserCallbackDispatcher+0x40d
02 00000000`0086e350 00007ffa`4e01fde4 wow64win!whcbClientThreadSetup+0x54
03 00000000`0086ed00 00000000`76ea1cbc ntdll!KiUserCallbackDispatcherContinue
04 00000000`0086ed58 00000000`76ea1c7b wow64cpu!CpupSyscallStub+0xc
05 00000000`0086ed60 00000000`76ea1199 wow64cpu!Thunk0Arg+0x5
06 00000000`0086ee10 00007ffa`4cfac77a wow64cpu!BTCpuSimulate+0x9
07 00000000`0086ee50 00007ffa`4cfac637 wow64!RunCpuSimulation+0xa
08 00000000`0086ee80 00007ffa`4dff18eb wow64!Wow64LdrpInitialize+0x127
09 00000000`0086f130 00007ffa`4dff17d3 ntdll!_LdrpInitialize+0xff
0a 00000000`0086f1d0 00007ffa`4dff177e ntdll!LdrpInitialize+0x3b
0b 00000000`0086f200 00000000`00000000 ntdll!LdrInitializeThunk+0xe
```
```
0:000> dg ss
                                                    P Si Gr Pr Lo
Sel        Base              Limit          Type    l ze an es ng Flags
---- ----------------- ----------------- ---------- - -- -- -- -- --------
002B 00000000`00000000 00000000`ffffffff Data RW Ac 3 Bg Pg P  Nl 00000cf3
```
On some occasions windows wow64 code will even use IRET for return into 32bit mode.

See here for some valuable results of reverse engineering WOW64:
[https://wbenny.github.io/2018/11/04/wow64-internals.html](https://wbenny.github.io/2018/11/04/wow64-internals.html).
